### ISSUE TITLE
Uncraftable trains from advtrains mod

### ISSUE THREAD

**Issue by Why-did-i** [pandorabox-io/pandorabox.io#608]
It would be cool if players could craft railbuses, japanese trains...

**Comment by BuckarooBanzay** [comment]
There should be some admin-shops that sell the missing trains and some other uncraftable stuff, did you take a look there?

The proper solution would be to add the recipes in the upstream mod (`advtrains`) but bringing in changes there involves either posting in the forums and pinging people there until someone notices or mailing the lead dev (orwell). None of that is really straight-forward :confounded:

**Comment by OgelGames** [comment]
The other option is to add the recipes to `pandorabox_custom` :wink:

**Comment by 6r1d** [comment]
How does this recipe sound like?

```lua
minetest.register_craft({
	output = 'advtrains:engine_railbus',
	recipe = {
		{'default:copperblock', 'default:torch', 'default:copperblock'},
		{'default:obsidian_glass', 'dye:red', 'default:obsidian_glass'},
		{'advtrains:wheel', 'advtrains:wheel', 'advtrains:wheel'},
	},
})
```

I hate to install torch, but IDK if advtrains has any kind of lamp.
I'd like to use `homedecor:glowlight_small_cube_14`, but it works only on Pandorabox and some other servers.
An alternative would be a `moreblocks:super_glow_glass`.

**Comment by 6r1d** [comment]
Umm, @Why-did-i, do you like it or not? :-)
I'd like to hear any comments about that.

**Comment by S-S-X** [comment]
@6r1d 
> How does this recipe sound like?

For me that sounds like way too simple, in my opinion recipes should use steam engine as baseline and switch/add just bit more expensive ingredient for fancy / modern engines or wagons.

For steam engine you need about 4 steel blocks, few basic items and crafting.
In my opinion for all train engines and wagons crafting should have similar steps that are required for steam engine.

That is, some basic engine parts as craft items, items could be some of these for example:
* Engine: maybe cast iron and some stuff from basic_materials
* Drive shaft: cast iron or stainless steel would probably be options
* Frame: probably steel blocks, maybe dye

Then combine those with glass, lights, etc. to get different engines. Maybe frame with similar decorative things + wheels could also be used to craft different wagons.
Idea would be to have basic train components to implement good recipes that require some effort (building train engine should not be too simple...) and be interesting and also would serve basic components for other engines/wagons if there's any.

**Comment by 6r1d** [comment]
There are two existing steam engines: `advtrains:detailed_steam_engine` and `advtrains:newlocomotive`, are you talking about them, @S-S-X?

If not, do you mean that we should introduce our own combination like `advtrains:base_engine`? A mix of `advtrains:boiler`, `basic_materials:heating_element` and `basic_materials:motor`, wheels and other stuff? If yes, how does that recipe for a base train engine sound?

```lua
minetest.register_craft({
	output = 'advtrains:base_engine',
	recipe = {
		{'advtrains:driver_cab', 'homedecor:glowlight_small_cube_14', 'technic:cast_iron_block'},
		{'basic_materials:heating_element', 'advtrains:boiler', 'basic_materials:motor'},
		{'advtrains:wheel', 'advtrains:wheel', 'advtrains:wheel'},
	},
})
```

**Comment by S-S-X** [comment]
> There are two existing steam engines: advtrains:detailed_steam_engine and advtrains:newlocomotive, are you talking about them, @S-S-X?

Yes.

> If not, do you mean that we should introduce our own combination like advtrains:base_engine

Almost there but not exactly what I meant. More like set of craft items, few basic train engine parts to introduce multi stage crafting and possibly alloying better materials (cast iron, possibly small amounts of stainless steel and brass) etc. if technic is available.

Then combining those with decorations would directly give you complete train engine (or wagon), similar how advtrains:detailed_steam_engine is crafted but with modern craft items made using modern materials for modern trains.

Like no base engine but parts for crafting complete train engines and wagons instead of using basic stuff like iron blocks, ingots, torches and such.

> A mix of advtrains:boiler, basic_materials:heating_element and basic_materials:motor, wheels and other stuff?

Probably boiler does does not belong to modern engines, for that I was thinking about adding few more generic parts like some of these: engine block, drive shaft, gearbox, generic frame (that could possibly also used for wagons). I think driver cab could be one of those parts yes.

Engine block:
![image](https://user-images.githubusercontent.com/50966843/106346440-d9f63900-62bf-11eb-9cd0-9d63469624ce.png)

Frame:
![image](https://user-images.githubusercontent.com/50966843/106346447-e2e70a80-62bf-11eb-8771-d372361d6f83.png)

Drive shaft:
![image](https://user-images.githubusercontent.com/50966843/106346571-ca2b2480-62c0-11eb-9aa4-904add360f7e.png)

Gearbox:
![image](https://user-images.githubusercontent.com/50966843/106346718-a5837c80-62c1-11eb-927e-c8ad773cf0ae.png)

**Comment by 6r1d** [comment]
That'd be nice. I haven't seen Naso for a while, though, so we might need to draw the icons.
I can try to do some in Blender, but the results may vary. :-)

**Comment by 6r1d** [comment]
A bit unrelated, but when I mentioned this discussion to sivarajan from Tunnellers Abyss, he had mentioned the [linetrack](https://gitlab.com/Hume2/linetrack) mod, adding an automated roadbus and boats.

**Comment by SivarajanSam** [comment]
Basically stuffs like the TCB, ATC, Lua atc Controller are Un-craftable to prevent normal players to use it  and it is of no use to them. But if there are players who build trains frequent you can add it to craft-able criteria for sure. On Tunnelers Abyss the craft recipe is not given. Instead the nodes that are not crafted and helpful for train programming and the train itself is sold. But it is best to have crafting recipe for all the nodes in adv trains. So I would support this as well.

**Comment by S-S-X** [comment]
> A bit unrelated, but when I mentioned this discussion to sivarajan from Tunnellers Abyss, he had mentioned the [linetrack](https://gitlab.com/Hume2/linetrack) mod, adding an automated roadbus and boats. ([Forum topic](https://forum.minetest.net/viewtopic.php?f=9&t=23528&hilit=linetrack).)

This mod also does not seem to have crafting recipes yet and even if it had it could also use many of same basic craft items that would be added for trains (engine block, gearbox, driveshaft). Probably it is just frame that would not be used for boat.. but maybe that could still be used for bus.

**Comment by 6r1d** [comment]
@S-S-X, would you like to add any other parts now? Send photos if yes.
I have updated the icons above. I can do some polishing and share Blend files. If it's not as important, let's talk about recipes more and add these to `pandorabox_custom` to test the process in-game.

If these icons will be useful, does CC0 work? Not sure if MIT is used for media files, but I'm ok with both.

**Comment by S-S-X** [comment]
> would you like to add any other parts now

Can't really think of any other good parts that would be needed right away, can be added later if/when someone has good idea for useful and generic enough additions.

> add these to pandorabox_custom to test the process in-game

I think we could also create new repository right away and add through pandorabox-mods after it is good enough, I think it would be better that way to reduce need for cleanup. Maybe add one under mt-mods organization?

> If these icons will be useful, does CC0 work? Not sure if MIT is used for media files, but I'm ok with both (and whatever Advtrains has, we want those to be merged back at a later date).

CC0 can work or CC-BY if you want to enforce attribution or CC-BY-SA if you want to also add relicensing restrictions. I'd be fine with all those, mostly dislike NC/ND like clauses because of high restrictions those put on usage and development.
Also personally I do not really like requirement to state changes as license clause for various reasons.
That said, it is your work and your choice.

Something like MIT for code and CC0 or CC-BY for media would probably be just fine.

**Comment by 6r1d** [comment]
> That said, it is your work and your choice.
I'm happy with CC0, I don't want any attribution for little drafts. (Thank you for clarifying about CC-BY-SA and other parts, but I think limiting the license won't help improve advtrains and improving the mod our goal, so there's no need.)

> Maybe add one under mt-mods organization?
Please do if you can.

**Comment by S-S-X** [comment]
Good and descriptive name for mod that provides these parts and recipes? Something like basic_machines? Good suggestions?

**Comment by OgelGames** [comment]
> Good and descriptive name for mod that provides these parts and recipes?

Well they are for crafting `advtrains` trains, so maybe something like `advtrains_crafts` or `advtrains_parts`?

**Comment by S-S-X** [comment]
> Well they are for crafting advtrains trains

And planned for boats, buses (those are still using advtrains) and being open for other things. However one idea was to not couple that too tightly to advtrains to encourage possible use with many other mods that add large machines or vehicles or just could use generic industrial machine parts and/or more advanced crafting for any machines.
That's also why trying to mostly introduce as generic machine parts as possible.

So basically copying idea of basic_materials mod but focusing on large machinery.

**Comment by OgelGames** [comment]
> However one idea was to not couple that too tightly to advtrains to encourage possible use with many other mods that add large machines or vehicles

Makes sense :+1: So something like `machine_parts` maybe?

**Comment by 6r1d** [comment]
Thanks, @S-S-X, now icons are [here](https://github.com/mt-mods/machine_parts/tree/master/icons) and the initial 3D models are [here](https://mega.nz/file/j40RGQhD#TBezmRiJWu__1srjQGRIGFJeeF-sjKL7MsX1im4T2jo) for anyone interested.

Time to decide on the recipes, then? :-)

(Also, should we maybe have an alternative to `homedecor:glowlight_small_cube_14` lamp inside a mod for lighting hulls? Something like `machine_parts:hull_light`?)

[![](https://i.imgur.com/wS5uOG5.png)](https://i.imgur.com/dTd5Ud9.png)

**Comment by S-S-X** [comment]
> Also, should we maybe have an alternative to homedecor:glowlight_small_cube_14 lamp inside a mod for lighting hulls?

Not sure.. is there some mods providing reasonable alternatives? Optional dependencies are thing and we could add recipe that has multiple fallback items, even all the way down to torch if needed.
Homedecor is one, for starters here's few possible homedecor:glowlight_small_cube_14, morelights_extras:f_block, morelights_modern:smallblock, morelights:bulb.

Not sure, kind of feels like lights do not really belong here and should be provided by some other mod and only used here.

**Comment by 6r1d** [comment]
> Not sure, kind of feels like lights do not really belong here and should be provided by some other mod and only used here.
Alright, let's not add them. Recipe code it is, then. :-)

**Comment by 6r1d** [comment]
A bunch of gears connected to one another (`machine_parts:gear_assembly`) sounds simple: gears, steel bar, steel strip and ingots as mounts / separators.

```lua
minetest.register_craft({
	output = 'machine_parts:gear_assembly',
	recipe = {
		{'basic_materials:gear_steel', 'default:steel_ingot', 'default:steel_ingot'},
		{'basic_materials:gear_steel', 'default:steel_ingot', 'default:steel_ingot'},
		{'basic_materials:gear_steel', 'basic_materials:steel_bar', 'basic_materials:steel_strip'}
	}
})
```

I'd argue we should make a gearbox a bit complex, but if anyone else wants to rebalance it, now is the time.

```lua
minetest.register_craft({
	output = 'machine_parts:gearbox',
	recipe = {
		{'machine_parts:gear_assembly', 'technic:stainless_steel_block', 'dye:violet'},
		{'machine_parts:gear_assembly', 'technic:stainless_steel_block', 'basic_materials:paraffin'},
		{'machine_parts:gear_assembly', 'basic_materials:steel_bar', 'basic_materials:steel_bar'}
	}
})
```

What about this engine block?
It is a bit on the expensive side and doesn't have a heating element, since I see little need for it, but looks useful to me.

```lua
minetest.register_craft({
	output = 'machine_parts:engine_block',
	recipe = {
		{'basic_materials:energy_crystal_simple', 'technic:stainless_steel_block', 'dye:green'},
		{'default:mese_crystal', 'technic:stainless_steel_block', 'basic_materials:copper_wire'},
		{'basic_materials:motor', 'basic_materials:motor', 'basic_materials:motor'}
	}
})
```

How about this driveshaft recipe?

```lua
minetest.register_craft({
	output = 'machine_parts:drive_shaft',
	recipe = {
		{'technic:stainless_steel_ingot', 'dye:yellow', 'technic:stainless_steel_ingot'},
		{'technic:stainless_steel_ingot', 'technic:stainless_steel_ingot', 'technic:stainless_steel_ingot'},
		{'technic:stainless_steel_ingot', 'basic_materials:oil_extract', 'technic:stainless_steel_ingot'}
	}
})
```

Maybe this wagon frame will work?

```lua
minetest.register_craft({
	output = 'machine_parts:wagon_frame',
	recipe = {
		{'technic:rubber', 'doors:trapdoor_steel', 'technic:rubber'},
		{'technic:stainless_steel_block', 'technic:stainless_steel_block', 'technic:stainless_steel_block'},
		{'technic:stainless_steel_ingot', 'technic:stainless_steel_ingot', 'technic:stainless_steel_ingot'}
	}
})
```

**Comment by Emil9736** [comment]
Good idea. Not only japans trains and rail-buses , other models from other countries of different time periods. Also electrical poles with contact electric grid, insulators, frames. But animations of trains is so laggy :/    
![Uploading Стойки опоры.JPG…]()
![f64_ligums_platforma_240714_011_2 (2)](https://user-images.githubusercontent.com/53637630/106641365-eec91a00-65a8-11eb-8c78-5ef3e153831d.jpg)

**Comment by Emil9736** [comment]
> 
> 
> @Emil9736, the first thing I see there is the sheer amount of the power lines.
> 
> Those are quite complex by themselves: different types of wires and cables to connect everything, insulators and towers. There's some "logic" on the ground to register trains and I am probably missing lots of stuff on this image.
> 
> For now, we need basic parts of the train itself. Can you see anything I or SX could've missed in terms of wagon / engine parts? SX posted some photos above.
> Look through recipes, it will be bad if anyone thinks they're too expensive or complex.

 We can simplify number of cables. For electric towers we can use ready mod bridger. For insulators  - easy and cheap recipes. We have already some logical units for train registration, rout control.  It can be released a bit latter. I think we should add driver stand with dashboard.

**Comment by S-S-X** [comment]
For recipes I think there should be more variation on materials, it feels that most of it is just stainless steel. To change this I did some simple research and here's some possible changes:
* Possibly some cast iron and/or carbon steel for engine_block and gearbox.
  For engine_block it would probably be fine to swap stainless steel blocks to cast iron blocks.
* For gearbox not sure actually but I think gearbox frame assembly probably is closer to cast iron than stainless steel.
  Gearbox could also use small amount of brass or even bronze.
* I think swapping some (not all of it) stainless steel with cast iron would be also good for wagon_frame.

Stainless steel for drive shaft is probably already closest available material (carbon steel would be another option) unless new materials (molybdenum for alloys https://en.wikipedia.org/wiki/41xx_steel) will be introduced, that however would probably belong to technic mod.

> Also electrical poles with contact electric power grid, insulators, frames.

This is probably a lot bigger change if implemented properly (meaning actual logic), I think we should probably add separate issue for this as reminder. Most of actual logic should probably be in advtrains mod, some of it possibly in technic mod but machine_parts could possibly also provide something. If nothing else then at least possibly tracking and developing this idea.

> I think we should add driver stand with dashboard.

Something like generic dashboard would be good craft item, this is generic part and very similar to ones used in boats, cranes, planes, harvesters and so on. It would provide good craft item with good reuse potential.

**Comment by 6r1d** [comment]
> For gearbox not sure actually but I think gearbox frame assembly probably is closer to cast iron than stainless steel.

> Gearbox could also use small amount of brass or even bronze.

Will this recipe work? :-)

```lua
minetest.register_craft({
	output = 'machine_parts:gearbox',
	recipe = {
		{'machine_parts:gear_assembly', 'technic:cast_iron_block', 'dye:violet'},
		{'machine_parts:gear_assembly', 'technic:cast_iron_block', 'basic_materials:paraffin'},
		{'machine_parts:gear_assembly', 'basic_materials:steel_bar', 'basic_materials:brass_ingot'}
	}
})
```

> some cast iron and/or carbon steel for engine_block

> For engine_block it would probably be fine to swap stainless steel blocks to cast iron blocks.

```lua
minetest.register_craft({
	output = 'machine_parts:engine_block',
	recipe = {
		{'basic_materials:energy_crystal_simple', 'technic:stainless_steel_block', 'dye:green'},
		{'default:mese_crystal', 'technic:cast_iron_block', 'basic_materials:copper_wire'},
		{'basic_materials:motor', 'basic_materials:motor', 'basic_materials:motor'}
	}
})
```

> I think swapping some (not all of it) stainless steel with cast iron would be also good for wagon_frame.

Will that work?

```lua
minetest.register_craft({
	output = 'machine_parts:wagon_frame',
	recipe = {
		{'technic:rubber', 'doors:trapdoor_steel', 'technic:rubber'},
		{'technic:stainless_steel_block', 'technic:cast_iron_block', 'technic:cast_iron_block'},
		{'technic:stainless_steel_ingot', 'technic:stainless_steel_ingot', 'technic:cast_iron_ingot'}
	}
})
```

> add separate issue for this as reminder

I see you've [added it already](https://github.com/mt-mods/machine_parts/issues/3).

> generic dashboard would be good craft item

Good idea. Now think you are a driver. What will you want to see? For me it's always "vitals" like speed first. They go on a separate display for reliability. I'd position this display on the right side, because this is "finishing" the image. Primary buttons can go there, too. (Assuming left-handed people will deal with it...)

Camera outputs go to the left, misc stuff with menus in the middle.

Some inspiration: [outdated one](https://thumbs.dreamstime.com/b/dashboard-rail-old-diesel-engine-passenger-train-61070334.jpg), [nicer one](https://www.thestatesman.com/wp-content/uploads/2019/02/Train-driver.jpg).

This recipe dependencies are overcomplicated, but convey what I want to convey with its construction well.
What will you change? When we will have a good recipe I'll add an icon.

```lua
minetest.register_craft({
	output = 'machine_parts:dashboard',
	recipe = {
		{'digiterms:lcd_monitor', 'digilines:lcd', 'digiterms:lcd_monitor'},
		{'mesecons_noteblock:noteblock', 'mesecons_luacontroller:luacontroller0000', 'mesecons_fpga:fpga0000'},
		{'digilines:wire_std_00000000', 'technic:rubber', 'technic:stainless_steel_ingot'}
	}
})
```

**Comment by 6r1d** [comment]
A new dashboard icon. I don't know if it "works" or not, but I picked some screenshots you can recognize and used a piece of Appolo screen.

[![](https://i.imgur.com/iyyJ3Dv.png)](https://i.imgur.com/0upmIMG.png)

**Comment by 6r1d** [comment]
Current list of dependencies is:

* `default`
* `basic_materials`
* `dye`
* `technic`
* `doors`
* `mesecons_noteblock`
* `mesecons_luacontroller`
* `mesecons_fpga`
* `digilines`
* `digiterms`

Now is a perfect time to veryfy recipes, dependencies and propose train recipes for the `railbus` and a `japanese train`, because later we'll have to change a larger codebase.

**Comment by TechDudie** [comment]
My craft suggestions

```
minetest.register_craft({
	output = 'machine_parts:gearbox',
	recipe = {
		{'machine_parts:gear_assembly', 'technic:cast_iron_block', 'machine_parts:gear_assembly''},
		{'machine_parts:gear_assembly', 'default:diamond_block', 'machine_parts:gear_assembly'},
		{'machine_parts:gear_assembly', 'basic_materials:steel_bar', 'machine_parts:gear_assembly''}
	}
})

minetest.register_craft({
	output = 'machine_parts:engine_block',
	recipe = {
		{'basic_materials:energy_crystal_simple', 'technic:stainless_steel_block', 'basic_materials:energy_crystal_simple'},
		{'basic_materials:copper_wire', 'technic:cast_iron_block', 'basic_materials:copper_wire'},
		{'basic_materials:motor', 'basic_materials:motor', 'basic_materials:motor'}
	}
})

minetest.register_craft({
	output = 'machine_parts:wagon_frame',
	recipe = {
		{'technic:rubber', 'default:mese_lamp', 'technic:rubber'},
		{'technic:stainless_steel_block', 'technic:cast_iron_block', 'technic:stainless_steel_block'},
		{'technic:stainless_steel_block', 'technic:stainless_steel_block', 'technic:stainless_steel_block'}
	}
})

minetest.register_craft({
	output = 'machine_parts:dashboard',
	recipe = {
		{'digilines:lcd', 'digilines:lcd', 'digilines:lcd'},
		{'mesecons_button:button_off', 'mesecons_luacontroller:luacontroller0000', 'mesecons_walllever:wall_lever_off'},
		{'digilines:wire_std_00000000', 'mesecons:wire_std_00000000', 'digilines:wire_std_00000000'}
	}
})

```

**Comment by 6r1d** [comment]
@TechDudie, thanks, now I'll answer you about the suggestions.
(I'm comparing recipes using [Meld](https://meldmerge.org/), tell me if I miss anything, please.)

First, `machine_parts:gearbox`. It is close to my idea before in regards to the use of stainless steel.
SX disagrees with that, chrome is expensive and costs resources.
You also add more `gear_assembly` blocks and a diamond block, which is on the expensive side of things. I was thinking my recipe is expensive at this point, I am sorry I haven't clarified that before.

Your variation of `machine_parts:engine_block` is not painted (I have included dye just for that), but has a mese block and additional energy crystal.
While I appreciate complexity (it makes game more fun when you solve hard problems), I can imagine people being unhappy with too expensive trains. I already heard something along the lines of that.

About `machine_parts:wagon_frame`: tell me, why do you replace a trapdoor (which tells there's a door in a recipe) with a lamp?

Finally, the dashboard. I used two LCD types to put extra attention: we have camera displays and "vitals" displays.
Camera display can be a high-resolution one, but "vitals" is more critical and should be simplified to [that level](https://hackaday.com/wp-content/uploads/2017/05/fran-blanche-el-display-thumb.jpg?w=600&h=600) for accessibility and reliability (imho).

**Comment by OgelGames** [comment]
> Current list of dependencies is:
> ...

That's quite a lot of dependencies, do you think maybe alternative recipes should be added to allow some of them to be optional? 🤔

**Comment by 6r1d** [comment]
> do you think maybe some alternative recipes should be added

Absolutely! It's a perfect moment to propose and discuss some.

**Comment by S-S-X** [comment]
> SX disagrees with that, chrome is expensive and costs resources.

I do disagree with that only because stainless steel is not actually used that much in real gearboxes, most of steel goes to frame assembly.

> ```
> minetest.register_craft({
> 	output = 'machine_parts:gearbox',
> 	recipe = {
> 		{'machine_parts:gear_assembly', 'technic:cast_iron_block', 'machine_parts:gear_assembly''},
> 		{'machine_parts:gear_assembly', 'default:diamond_block', 'machine_parts:gear_assembly'},
> 		{'machine_parts:gear_assembly', 'basic_materials:steel_bar', 'machine_parts:gear_assembly''}
> 	}
> })
> ```

And for same reason I would disagree with suggestion to use `default:diamond_block` in gearbox recipe, how much diamonds you'll find from actual gearbox? I did however suggest adding more expensive recipes through SAE steel grades, here's simple information about composition and use: https://en.wikipedia.org/wiki/SAE_steel_grades

However, adding new ores (molybdenum mostly) and new metal alloys is good idea only if there's more use for those, this would require tight cooperation at least with technic mod. At minimum some plans for new stuff or recipe updates that would use those new materials.

> ```
> minetest.register_craft({
>	output = 'machine_parts:engine_block',
>	recipe = {
>		{'basic_materials:energy_crystal_simple', 'technic:stainless_steel_block', 'basic_materials:energy_crystal_simple'},
>		{'basic_materials:copper_wire', 'technic:cast_iron_block', 'basic_materials:copper_wire'},
>		{'basic_materials:motor', 'basic_materials:motor', 'basic_materials:motor'}
>	}
> })
> ```

And again, what is `basic_materials:energy_crystal_simple` actually doing in engine_block? I think it is supposed to be energy storage device.
Here also stainless steel is not actually really dominating material in actual engine blocks, fully assembled engine block can contain some stainless steel but much less than either cast iron, or some grade of carbon steel.

For suggested `machine_parts:wagon_frame` recipes, I think any of those are okay and here we can also balance costs of end products while keeping recipes somewhat realistic. Frame materials can vary a lot more compared to engine / powertrain parts.

For `machine_parts:dashboard`, this recipe can also vary a lot more so just have to pick good enough recipe with least dependencies. There probably should however be fallback recipe that does not use any mesecons or digiline parts.

**Comment by S-S-X** [comment]
> > Current list of dependencies is:
> > ...
> 
> That's quite a lot of dependencies, do you think maybe some alternative recipes should be added to allow some of them to be optional? 🤔

* default, probably has to be hard dependency.
* basic_materials, probably also good to have as hard dependency. This is also fairly common dependency for some popular mods.
* dye, this probably can be optional and if not available replaced with something simple from default.
* technic, this can be optional dependency by replacing advanced alloys with default metal ingots / blocks.
* doors, I think this is not very and can be optional dependency replacing doors with some very basic things if not available.
* mesecons_noteblock, maybe this should not be used at all in recipes or could also be optional.
* mesecons_luacontroller, this can be optional dependency and use fallback route like Luacontroller -> CLU -> Simple IC -> Gold block
* mesecons_fpga, maybe this should not be used at all in recipes or could also be optional.
* digilines, this must be optional unless digiline logic will be actually used. Fallback something like Luacontroller.
* digiterms, this must be optional as these will be only for crafting, fallback route similar to Luacontroller

That would be 2 hard dependencies and 7 optional dependencies where few optionals wont affect crafting very much.

**Comment by 6r1d** [comment]
> doors, I think this is not very and can be optional dependency replacing doors with some very basic things if not available.
yeah, my motivation there was `advtrains:boiler` recipe, it has a door as a dependency. I'd be happy to completely remove the said door, as well.

> basic_materials, probably also good to have as hard dependency
good, because a copper wire is a wire, works as good as digilines for me, especially with plastic sheets and rubber in a recipe ;-)

> technic, this can be optional dependency by replacing advanced alloys with default metal ingots / blocks.
Well, if people can avoid alloys, they will, so I'm not sure it should be optional. If technic is available, alloys should be used.
CLU is a part of Technic, too.

> dye
> something simple from default
bones and grass?

**Comment by TechDudie** [comment]
@S-S-X, all my diamond block uses are partially to make the trains expensive, but also because diamond, well ofc, is very durable and strong.

**Comment by S-S-X** [comment]
> but also because diamond, well ofc, is very durable and strong.

Just saying that you have to do some more research on material properties... toughness of diamond is very (extremely!) poor compared to any of automotive steels.

**Comment by 6r1d** [comment]
> toughness of diamond is very (extremely!) poor compared to any of automotive steels.

Agreed, @S-S-X

But let's address this aspect:

> all my diamond block uses are partially to make the trains expensive

Why "*partially*", @TechDudie? Do they have a construction use, as a part? Or, instead, as "*make this thing strong, because diamonds are*"?

If we check current cost of a japanese train, it's 100 mg.
We should not make it more expensive or players will be unhappy. A bit less expensive might be ok. **Coil** sells diamond blocks for 190 mg and a bit of difference begins here. He is a reliable person to tell us the costs of stuff, he counts each.

Now let's think about two approaches to content games take. By content I mean graphics, sounds and so on.

---

One approach is "**leveling up replaces content**". You go from wooden sword to lava sword, and from that position you can only laugh at a boring wooden thing. Some things are "more magical", some things are "more mundane". It doesn't always work when you have limits on content and mobile users aren't happy with RAM being overloaded.

Another approach is "**each thing has its strong and weak sides**". Little example we have is a dough recipe. Beer contains yeast, which is used in cooking. Baking flour in-game currently gives you bread, but baking some flour with beer, sugar and salt gives you more. Just look at the possibilities in [Little Alchemy](https://littlealchemy.com/)! The risk here is to overcomplicate components and get beyond 9 ones in a craft menu: then it'll require new content for intermediate parts.

I'd say I like the second approach more: we haven't exhausted possible uses for in-game textures and sounds yet and it's fun in general.

**Comment by S-S-X** [comment]
If we need more expensive and interesting recipes and more content to game I would lean towards actually adding molybdenum ore and alloys I've been talking about, see following article for easy to read generic all around info:
https://www.britannica.com/technology/molybdenum-processing

That could also be first ore that could require HV furnaces for smelting because of very high melting point, a lot higher than other metals currently available.
Some of current technic recipes could be updated to use alloys requiring molybdenum and it would serve as good ingredient for some more or less planned upgrades.

That would add content on many levels, would be fairly big content upgrade for mt-mods technic with relatively simple changes to actual code and would also make it stand out with more interesting generic content.

I really started thinking this should be considered as real option.

**Comment by TechDudie** [comment]
Well @S-X-X, I must have accidentally forgot a part. I meant partially expensive, partially strong.

**Comment by S-S-X** [comment]
Can be closed by https://github.com/pandorabox-io/pandorabox-mods/pull/1554

**Comment by 6r1d** [comment]
What needs to be done now, @S-S-X? Merging into the game?

**Comment by S-S-X** [comment]
> What needs to be done now, @S-S-X? Merging into the game?

Merge pandorabox-io/pandorabox-mods#1554 is all that needs to be done to get usable recipes.

Other things nice to have would be these but not strictly required to allow crafting trains and wagons:
* https://github.com/mt-mods/machine_parts/issues/9
* https://github.com/mt-mods/machine_parts/issues/10

**Comment by S-S-X** [comment]
Recipes been available for almost few weeks already, closing this.
